### PR TITLE
chore: validate geopackage index

### PR DESCRIPTION
**Pull request by CL-SHLOMIKONCHA**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✔/✖                                                                        |
| New feature     | ✔/✖                                                                        |
| Breaking change | ✔/✖                                                                        |
| Deprecations    | ✔/✖                                                                        |
| Documentation   | ✔/✖                                                                        |
| Tests added     | ✔/✖                                                                        |
| Chore            | ✔/✖                                                                       |

Related issues: #XXX , #XXX ...
Closes #XXX ...

Further  information:

### CODE REVIEW

File: src/layers/models/fileValidator.ts [hunk 1]
```diff
@@ -24,6 +26,24 @@ export class FileValidator {
   }
 
   public validateGpkgFiles(files: string[]): boolean {
+    this.validateGpkgIndex(files);
+
+    const allValid = this.validateGpkgExtension(files);
```
Comment (asafMasa) [inline src/layers/models/fileValidator.ts:31]: validate extension should be called before "validateGpkgIndex" 

const isExtensionValid = this.validateGpkgExtension(files);
if (!isExtensionValid) {
return false }
... validateGpkgIndex()

File: src/layers/models/layersManager.ts [hunk 2]
```diff
@@ -46,10 +46,10 @@ export class LayersManager {
     if (convertedData.id !== undefined) {
       throw new BadRequestError(`received invalid field id`);
     }
+    await this.validateFiles(data);
     const jobType = await this.getJobType(data);
     const taskType = this.getTaskType(jobType, data.fileNames);
 
-    await this.validateFiles(data);
     await this.validateJobNotRunning(resourceId, productType);
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:53]: move the "validateJobNotRunning" method before the "getJobType" function

File: src/serviceClients/sqliteClient.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  private readonly db: SQLiteDB;
+
+  public constructor(packageName: string) {
+    this.logger = container.resolve(Services.LOGGER);
+    this.config = container.resolve(Services.CONFIG);
+    this.layerSourcesPath = this.config.get<string>('layerSourceDir');
+    this.packageName = packageName;
+    this.fullPath = join(this.layerSourcesPath, this.packageName);
+    this.packageNameWithoutExtension = this.packageName.substring(0, this.packageName.indexOf('.'));
+    this.db = new Database(this.fullPath, { fileMustExist: true });
```
Comment (asafMasa) [inline src/serviceClients/sqliteClient.ts:24]: this row opens the DB connection - https://github.com/WiseLibs/better-sqlite3/blob/HEAD/docs/api.md#new-databasepath-options

you should move it inside the "getGpkgIndex()" method

File: src/serviceClients/sqliteClient.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+  public getGpkgIndex(): unknown {
+    const sql = `SELECT * 
+                  FROM sqlite_master 
+                    WHERE type = 'index' AND tbl_name='${this.packageNameWithoutExtension}' AND sql LIKE '%zoom_level%'
+                     AND sql LIKE '%tile_column%'
+                      AND sql LIKE '%tile_row%';`;
+
+    const index = this.runStatement(sql);
+    this.closeConnection();
```
Comment (asafMasa) [inline src/serviceClients/sqliteClient.ts:35]: add try, finally to the method content, in the finally close the db connection

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: tests/unit/layers/models/fileValidator.spec.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    jest.clearAllMocks();
+    jest.restoreAllMocks();
+    clearMockConfig();
+    initMockConfig();
+  });
+
+  describe('validateGpkgIndex', () => {
+    it('should fail if geopackage does not have a tiles index', function () {
+      setValue({ layerSourceDir: 'tests/mocks/files' });
+      const testData: string[] = ['no-index.gpkg'];
```
Comment (asafMasa) [inline tests/unit/layers/models/fileValidator.spec.ts:22]: you can create the index with mock and add a test for existing index (valid response)

a file with invalid GPKG can be added to the integration tests

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: src/layers/models/fileValidator.ts (hunk 1 above)
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/fileValidator.ts:31]: fixed

File: src/layers/models/layersManager.ts (hunk 2 above)
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/layersManager.ts:53]: fixed

File: src/serviceClients/sqliteClient.ts (hunk 4 above)
Comment (CL-SHLOMIKONCHA) [reply src/serviceClients/sqliteClient.ts:35]: fixed

File: src/serviceClients/sqliteClient.ts (hunk 3 above)
Comment (CL-SHLOMIKONCHA) [reply src/serviceClients/sqliteClient.ts]: fixed

File: tests/unit/layers/models/fileValidator.spec.ts (hunk 5 above)
Comment (CL-SHLOMIKONCHA) [reply tests/unit/layers/models/fileValidator.spec.ts]: fixed - added integration tests

File: tests/unit/layers/models/fileValidator.spec.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    clearMockConfig();
+    initMockConfig();
+  });
+
+  describe('validateGpkgIndex', () => {
+    it('should fail if geopackage does not have a tiles index', function () {
+      const getGpkgIndexFn = jest.spyOn(SQLiteClient.prototype as unknown as { getGpkgIndex: () => unknown }, 'getGpkgIndex');
+      getGpkgIndexFn.mockReturnValue(undefined);
+      setValue({ layerSourceDir: 'tests/mocks' });
+      const testData: string[] = ['unndexed.gpkg'];
```
Comment (asafMasa) [inline tests/unit/layers/models/fileValidator.spec.ts:27]: spelling, remove no-index

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: src/serviceClients/sqliteClient.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    this.logger = container.resolve(Services.LOGGER);
+    this.config = container.resolve(Services.CONFIG);
+    this.layerSourcesPath = this.config.get<string>('layerSourceDir');
+    this.packageName = packageName;
+    this.fullPath = join(this.layerSourcesPath, originDirectory, this.packageName);
+    this.packageNameWithoutExtension = this.packageName.substring(0, this.packageName.indexOf('.'));
+  }
+
+  public getGpkgIndex(): unknown {
+    const db = new Database(this.fullPath, { fileMustExist: true });
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/serviceClients/sqliteClient.ts]: fixed

File: tests/unit/layers/models/fileValidator.spec.ts (hunk 6 above)
Comment (CL-SHLOMIKONCHA) [reply tests/unit/layers/models/fileValidator.spec.ts]: fixed

File: src/serviceClients/sqliteClient.ts [hunk 8]
```diff
@@ -23,8 +24,8 @@ export class SQLiteClient {
   }
 
   public getGpkgIndex(): unknown {
-    const db = new Database(this.fullPath, { fileMustExist: true });
     try {
+      this.db = new Database(this.fullPath, { fileMustExist: true });
```
Comment (asafMasa) [inline src/serviceClients/sqliteClient.ts:28]: no need for it to be a class member
let db?: SQLiteDB;
try { 
db = new Database(this.fullPath, { fileMustExist: true });
...
}  finally {
 if(db) {
        db.close();
      }
}

**Review by asafMasa — CHANGES_REQUESTED** [review]
fix lint

**Review by asafMasa — APPROVED** [review]

File: src/serviceClients/sqliteClient.ts (hunk 8 above)
Comment (CL-SHLOMIKONCHA) [reply src/serviceClients/sqliteClient.ts]: fixed

**Pull request merged**